### PR TITLE
persist: implement stats trimming

### PR DESCRIPTION
**Pull request by danhhz**
Given a budget, we do a couple rounds of making stats smaller, until it fits within a given budget. These are applied in order of how lossy they are (i.e. how likely in practice to increase the false positive rate).

Touches #12684 

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by bkirwi — APPROVED** [review]
Great!

File: src/persist-types/src/stats.rs [hunk 1]
```diff
@@ -61,6 +62,25 @@ pub trait DynStats: std::fmt::Debug + Send + Sync + 'static {
     fn into_proto(&self) -> ProtoDynStats;
 }
 
+mod private {
+    /// Statistics serialization costs
+    pub trait StatsCost {
+        /// A proxy for the cost to serialize these stats, in bytes.
+        ///
+        /// WIP should we tie this to some specific meaning: e.g. the size of
```
Comment (danhhz) [inline src/persist-types/src/stats.rs:70]: @bkirwi any thoughts on this?

File: src/persist-types/src/stats.rs [hunk 2]
```diff
@@ -310,9 +376,140 @@ mod impls {
...
+                JsonStats::Maps(x) => x.values().map(|x| x.cost()).sum(),
+            }
+        }
+        fn trim(&mut self) {
+            match self {
+                JsonStats::None
+                | JsonStats::Mixed
+                | JsonStats::JsonNulls
+                | JsonStats::Bools(_)
+                | JsonStats::Strings(_)
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist-types/src/stats.rs:497]: oh yeah, good point

File: src/persist-types/src/stats.rs (hunk 1 above)
Comment (bkirwi) [reply src/persist-types/src/stats.rs:70]: Agree with everything you wrote in the comment, basically. I'm fine with defining this pretty loosely for now, and after the next round of testing we should have a better idea of how this behaves in real life / whether it's worth investing the time to make this more strict / predictable.

Comment (danhhz) [reply src/persist-types/src/stats.rs]: 👍 swapped it to TODO(mfp) so we don't forget to come back to it

**Comment by danhhz** [comment]
TFTR!

**Pull request merged**